Add new Stripe payment method compatible with the new Stripe Payment Intents API

luisramos0 opened the pull request:
#### What? Why?

Relates to #4180
Closes #4686 and #4173

The decision to duplicate the existing Stripe payment method has the objective of making the migration as smooth as possible and also of making it possible for us, after we migrate all users, to delete the original stripe payment method (and its code), and keep only the new one (that uses the new SCA compatible APIs).
Although this new API was made by stripe for European SCA it does work worldwide, we don't need to keep both integrations :+1:

This PR is not introducing any new auth processes yet. It should just work like the existing stripe payment method.

I have added entries to rubocop_manual_todo for the remaining rubocop issues. Code climate is not picking the exceptions list from rubocop_manual_todo...

In terms of tests we are not adding any feature tests here as we plan to get something new with #4639

#### What should we test?
We need to verify that the existing Stripe payment method works.
We need to verify the new payment method, it should just work like the existing stripe payment method.
Processes we need to test:
1. make a payment without saving card
2. make a payment saving card
3. make a payment using previously saved card
4. add a card on account section and use it to checkout
5. delete saved card
6. capture the payment in the admin payments section of an order
7. as a user authorize a hub to use your card and verify the subscription payments work with that card

The following details are valid before #4719, if we are testing 4719, all cases above should work.
<details>
As discussed in comments below: case 4 above is not working yet for this payment method. But case 7, subscriptions, is working.
Because case 4 doesnt work (if you add a card in the account section and use it in a subscription with this new payment methods, the payments will fail). To add a card that works for a subscription with this new payment method you need to checkout an order with this new payment method with the save card option enabled. After that go to account to authorize the card you have used to checkout.
</details>

#### Release notes
Changelog Category: Added
Added new payment method that integrates with the new Stripe Payment Intents API that is compatible with SCA. This is a initial preparation for us to implement the more complex payment authentication flows that banks will start to enforce in Europe soon.

Matt-Yorkley commented:
Nice work! Looks like some of the stubbed requests in the new specs need adjusting a bit...

luisramos0 commented:
yes, thanks @Matt-Yorkley I am finishing that spec today. I wanted to get some early feedback :+1:

luisramos0 commented on lib/stripe/profile_storer.rb:
```diff
@@ -4,17 +4,22 @@
 
       if response.success?
-        attrs = source_attrs_from(response)
+        if response.params['customer'] # Payment Intents API
```
yes, I did that in a later commit :+1:

luisramos0 commented on lib/stripe/profile_storer.rb:
```diff
@@ -24,11 +29,13 @@ def create_customer_from_token
+                  login: Stripe.api_key,
+                  address: address_for(@payment)
+                }
+      options = options.merge({ stripe_account: @stripe_account_id }) if @stripe_account_id.present?
```
yeah, nice. I added this.

luisramos0 commented:
ok, I rebased and added the two last commits: make one method better looking and add some missing translations.

There are two test scenarios that are not working yet. 4 and 7.
4. Adding a card in the account section
the problem here is that this customer form does not have any association with a payment method, so we dont know if this will be used for a stripe payment method with SCA or the legacy stripe api. Currently the form is adding the card to the legacy stripe api and these cards will only work with payment methods that use the legacy stripe api (non sca). I am not sure how we should proceed here because while the new sca payment method is not completed and all user data migrated, we will still want the old stripe payment methods to work.

7. subscriptions
Without point 4 above cards can only be added through checking out with save card option.
There are a few more things to be done on the subscriptions front to make this work.

I think we should ship this PR as is if no one disagress. the new stripe sca payment method will be available but only works in limited list of use cases. We can then improve it until we can replace the old one.

lin-d-hop commented:
Thanks @luisramos0 I agree with your approach above. 

Do you foresee any issues when we come to merge the old API user data with the new API user data. As the above approach means we'll have split user data including potentially the same user across both APIs. I guess this is totally handled by Stripe?

RachL commented:
Something I'm not sure to understand: if we upgrade the API, doesn't it means that we should disable the current form that saves card in the legacy API?
I mean, migration is always a mess, so the less card to migrate the better?

luisramos0 commented:
@RachL this is not related to upgrading to the latest api version.
this is related to using a different api: charges api (non sca compliant) vs payment intents api (sca compliant).
All stripe api versions include the charges api but we want to upgrade so that we get the recent versions of the payment intents api with all the sca features.

We can have both payment methods work at the same time, we just need to see what to do with this customer form.

if you agree we can just go forward with this PR as is. I can start investigate the easiest approach afterwards:
- maybe I can make the add card form add to both apis
- maybe I can make the cards added through the old api (charges api) work with the new api (payment intents api) - this would be great because we wouldnt need to migrate the user cards
- other approach

luisramos0 commented:
@lin-d-hop answering your question: I still have to see if I can use the existing data in the new api. I havent managed so far but I think it will be possible.

Conclusion: lets go forward with this PR that adds the new payment method (Stripe SCA) and works with the new payment intents api. This way we can already validate it in staging and live. And meanwhile we can work on making this payment method ready for the full process.

luisramos0 commented:
ok, it looks like subscriptions are working :-)

This is ready to be reviewed and tested :+1:
@mkllnk I have added the 4 last commits since your review.
@Matt-Yorkley can you please review?
Thanks!!!!

luisramos0 commented:
1. I dont think this will be a problem because the new payment method is basically working. The only case not working is for users to add a card in the account section. Frontoffice. And they will see it doesnt work because the payments will fail. Also I think we will be ready very soon and we will probably rename the existing stripe one to something like "Stripe (legacy)". Having said that: what do other people think? Shall we name this new payment method "Stripe SCA (beta)"?

2. "a card added with the old API is not usable in the intents API " this is not a fact and probably it is incorrect, I need to investigate further.
It is legal to add a card as an admin, yes, for example, buying by phone. what we will have is a process to send the customer an email to authorize the transaction if that is what the bank requires. These cases are covered in the stripe epic issues if you want to see more details: https://github.com/openfoodfoundation/openfoodnetwork/issues/4170

Matt-Yorkley commented on app/assets/javascripts/admin/payments/services/payment.js.coffee:
```diff
@@ -30,11 +30,22 @@ angular.module('admin.payments').factory 'Payment', (AdminStripeElements, curren
               month: @form_data.card.exp_month
               year: @form_data.card.exp_year
           }
+        when 'stripe_sca'
```
Is duplicating this block necessary? It's identical to the one above.

luisramos0 replied:
I dont know the exact syntax for the when with two cases...

but actually, this answers to the other comments below: these changes are optimized for deletion of the stripe connect method very soon.

do you mind if I leave my version, basically separate lines for each method? it will be straight forward to delete them soon.

Matt-Yorkley commented on app/controllers/spree/admin/payment_methods_controller.rb:
```diff
@@ -134,19 +134,29 @@ def load_hubs
       end
+
+      def stripe_payment_method?
+        @payment_method.try(:type) == "Spree::Gateway::StripeConnect" ||
```
`["Spree:Gateway::StripeConnect", "Spree::Gateway::StripeSCA"].include? @payment_method.try(:type)` ?

luisramos0 replied:
:+1:

Matt-Yorkley commented on app/controllers/spree/admin/payment_methods_controller.rb:
```diff
@@ -134,19 +134,29 @@ def load_hubs
+      end
+
+      def stripe_provider?(provider)
+        provider.name.ends_with?("StripeConnect") ||
```
`provider.name.ends_with?("StripeConnect", "StripeSCA")` ?

luisramos0 replied:
:+1:

Matt-Yorkley commented on lib/open_food_network/subscription_payment_updater.rb:
```diff
@@ -30,7 +30,8 @@ def create_payment
 
     def card_required?
-      payment.payment_method.is_a? Spree::Gateway::StripeConnect
+      payment.payment_method.is_a?(Spree::Gateway::StripeConnect) ||
```
`[Stripe::Gateway::StripeConnect, Spree::Gateway::StripeSCA].include? payment.payment_method.class` ?

Matt-Yorkley replied:
There are several of these checks that all look very similar. Maybe extracting it might be nice at some point?

luisramos0 replied:
we will remove the existing stripe method soon so it will become what it was before...
but it would be good to have a central place where this check is done yes, I thought it would take a little bit to do it and I decided to skip it for now...

Matt-Yorkley commented on spec/requests/checkout/stripe_sca_spec.rb:
```diff
@@ -0,0 +1,281 @@
+        end
+      end
+
+      context "when the payment_metho request returns an error message" do
```
`payment_metho` :trollface:

luisramos0 replied:
fixed :+1:

Matt-Yorkley commented on spec/requests/checkout/stripe_sca_spec.rb:
```diff
@@ -0,0 +1,281 @@
+require 'spec_helper'
+
+describe "checking out an order with a Stripe SCA payment method", type: :request do
```
:heart:

Matt-Yorkley commented on app/models/spree/gateway/stripe_connect.rb:
```diff
@@ -9,12 +9,6 @@ class StripeConnect < Gateway
 
       attr_accessible :preferred_enterprise_id
 
-      CARD_TYPE_MAPPING = {
```
Isn't this needed anywhere for Stripe Connect? Are these changes to the old Stripe Connect code going to break it?

luisramos0 replied:
no, when adapting the new method I found this is dead code in the existing method.
in the commit message for this deletion you can see the commit when the code became dead:
https://github.com/openfoodfoundation/openfoodnetwork/pull/4672/commits/785e89624cf02b084afd26cb3c9e116d4252664c

luisramos0 commented:
thanks for the review Matt.

Yeah, we can change its name or make it super admin only.
I am pretty confident we will make this migration really quickly so I'd just go for it as is.

luisramos0 commented:
I managed to have 2 typos in this last commit.
I fixed that last commit and also rebased.

luisramos0 commented:
I am blocking this PR with #4719.

There's no point in testing this PR and then test 4719 again. 4719 implements a different logic to store cards (puts them in the stripe platform account of the instance and then clones them to connected stripe accounts of the seller).

Let's wait for 4719 to be ready for testing and test it instead (it includes this PR code).

luisramos0 commented:
rebased to resolve conflicts and pushed to upstream so that this is also seen in #4719

luisramos0 commented:
tested as part of 4719, merging.
